Source/DafnyCore/Rewriters/RefinementTransformer.cs

```diff
@@ -401,31 +402,31 @@ public RefinementTransformer(Program p)
         nwn.Constraint = dn.Constraint == null ? null : refinementCloner.CloneExpr(dn.Constraint);
         nwn.WitnessKind = dn.WitnessKind;
         nwn.Witness = dn.Witness == null ? null : refinementCloner.CloneExpr(dn.Witness);
-        m.TopLevelDecls[index] = MergeClass((NewtypeDecl)nw, (NewtypeDecl)d);
+        nwPointer.Set(MergeClass((NewtypeDecl)nw, (NewtypeDecl)d));
       } else if (nw is NewtypeDecl) {
         // `.Basetype` will be set in AddDefaultBaseTypeToUnresolvedNewtypes
         Reporter.Error(MessageSource.RefinementTransformer, nw, commonMsg, nw.WhatKind, nw.Name);
       } else if (nw is IteratorDecl) {
         if (d is IteratorDecl) {
-          m.TopLevelDecls[index] = MergeIterator((IteratorDecl)nw, (IteratorDecl)d);
+          nwPointer.Set(MergeIterator((IteratorDecl)nw, (IteratorDecl)d));
         } else {
           Reporter.Error(MessageSource.RefinementTransformer, nw, "an iterator declaration ({0}) in a refining module cannot replace a different kind of declaration in the refinement base", nw.Name);
         }
       } else if (nw is TraitDecl) {
         if (d is TraitDecl) {
-          m.TopLevelDecls[index] = MergeClass((TraitDecl)nw, (TraitDecl)d);
+          nwPointer.Set(MergeClass((TraitDecl)nw, (TraitDecl)d));
         } else {
           Reporter.Error(MessageSource.RefinementTransformer, nw, commonMsg, nw.WhatKind, nw.Name);
         }
       } else if (nw is ClassDecl) {
         if (d is ClassDecl) {
-          m.TopLevelDecls[index] = MergeClass((ClassDecl)nw, (ClassDecl)d);
+          nwPointer.Set(MergeClass((ClassDecl)nw, (ClassDecl)d));
         } else {
           Reporter.Error(MessageSource.RefinementTransformer, nw, commonMsg, nw.WhatKind, nw.Name);
         }
       } else if (nw is DefaultClassDecl) {
         if (d is DefaultClassDecl) {
-          m.TopLevelDecls[index] = MergeClass((DefaultClassDecl)nw, (DefaultClassDecl)d);
+          m.DefaultClass = (DefaultClassDecl)MergeClass((DefaultClassDecl)nw, (DefaultClassDecl)d);
```

Comment:
I'm not very familiar with this part of the code, but is it not necessary to `nwPointer.Set(m.DefaultClass)` here?

Reply:
This will work as well, since there's only one `DefaultClass` field so assigning to it will do the same as using the `Set` of that field's pointer, but what you suggest looks better. I will change that in the follow-up PR.